### PR TITLE
Factory refactor + testing, define Locale and Provider constants

### PR DESCRIPTION
**Pull request by pr0ggy**
Wanted to make **`Faker\Factory`** a bit more SRP-friendly...broke out separate **`Locale`** objects and associated factory, as well as an object to handle locale qualification.  Full unit tests included ([code coverage report](https://github.com/fzaninotto/Faker/files/73583/code-coverage.zip) attached).  These are injected into the **`Faker\Factory`** via a new `init.php`, required from `Faker/Factory.php`.  Also pulled Locale and Provider ID constants out into `const.php`.  Unit tests make use of the Mockery engine, which has been included in `composer.json`.

No actual functionality changes should result, Faker creation and usage is exactly as before.  Ran multi-thousand iteration comparison runs, no appreciable performance impact.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 vendor
 composer.lock
+code-coverage
```
Comment (EmanueleMinotto) [inline .gitignore]: Could you add a newline here? Reasons are listed here: http://stackoverflow.com/q/729692/871913

File: composer.json [hunk 2]
```diff
@@ -15,7 +15,8 @@
     "require-dev": {
         "phpunit/phpunit": "~4.0",
         "squizlabs/php_codesniffer": "~1.5",
-        "ext-intl": "*"
+        "ext-intl": "*",
+        "Mockery/Mockery": "^0.9.4"
```
Comment (EmanueleMinotto) [inline composer.json]: This could be all lowercase -> `mockery/mockery`

**Comment by pr0ggy** [comment]
Bah forgot to squash the last merge, will correct

File: phpunit.xml.dist [hunk 3]
```diff
@@ -1,12 +1,24 @@
...
+
+<phpunit
+    bootstrap="testing_bootstrap.php"
+    colors="true"
+>
+
+    <filter>
+      <whitelist processUncoveredFilesFromWhitelist="true">
+        <directory suffix=".php">./src/Faker/</directory>
+        <!-- <file>/path/to/file</file>
```
Comment (fzaninotto) [inline phpunit.xml.dist]: do not commit commented out code

**Comment by fzaninotto** [comment]
Sorry, I fail to see that as an improvement. I know you put a lot of effort into it, but I especially fly away from overengineering.

Feel free to reopen if you think you can illustrate a significant improvement in usage for Faker (not in code style, which is subjective).

**Pull request closed without merging**